Automattic/pocket-casts-ios
# End of Year: add stories Data Source and gestures

| 📘 Project: #376 |
|:---:|

This PR adds a data source protocol to populate the Stories view with stories and also adds a few gestures.

https://user-images.githubusercontent.com/7040243/195642253-8da709b6-d2da-43ef-8bdb-743f727fbe45.mp4

## About the data source

I tried to be as generic as possible with how we display stories to allow different cases to be used without much effort.

In order to show stories, we need to pass a data source that conforms to `StoriesDataSource`. Each story is simply a `View`.

This makes it possible for a story to be a custom view, an image, or a video. It's also generic enough to be used for other purposes in the future.

Take a look at the `TestStoriesDataSource` for an example.

The overall API is still in its early days, so we'll change it as the feature gets developed.

## To test

1. Enable the `endOfYear` flag in `FeatureFlag.swift`
2. Run the app
3. When the prompt appears, tap "View My 2022"
4. ✅ Check that each story is shown for 5s and that the story indicator (the rectangles at the top) are reflecting the state
5. ✅ Tap and hold, check that the current story is "paused". The rectangle stop progressing and the story doesn't change
6. ✅ While seeing the yellow story, tap the left portion of the screen, you should go back to the purple one
7. ✅ While seeing the purple story, tap the right portion of the screen, you should go to the next story (yellow)
8. ✅ Swipe down and make sure the view is dismissed ¶

¶ Swipe down can be improved with better effects (like with the view following the gesture) but this will be done in another task

## Checklist
